File: nodejs/lesson02/solution/hello.js
const assert = require("assert");
const { initTracer } = require("../../lib/tracing");

const tracer = initTracer("hello-world");

const sayHello = helloTo => {
  const span = tracer.startSpan("say-hello");
  const ctx = { span };
  span.setTag("hello-to", helloTo);
  const helloStr = formatString(ctx, helloTo);
  printHello(ctx, helloStr);
  span.finish();
};

const formatString = (ctx, helloTo) => {
  ctx = {
    span: tracer.startSpan("formatString", { childOf: ctx.span }),
  };
  const helloStr = `Hello, ${helloTo}!`;
  ctx.span.log({
    event: "string-format",
    value: helloStr,
  });
  ctx.span.finish();
  return helloStr;
};

const printHello = (ctx, helloStr) => {
  ctx = {
    span: tracer.startSpan("printHello", { childOf: ctx.span }),
  };
  console.log(helloStr);
  ctx.span.log({ event: "print-string" });
  ctx.span.finish();
};

assert(process.argv.length == 3, "Expecting one argument");
const helloTo = process.argv[2];

sayHello(helloTo);

tracer.close(() => process.exit());

File: python/lesson02/solution/hello.py
import sys
import time
from lib.tracing import init_tracer

def say_hello(hello_to):
    with tracer.start_active_span('say-hello') as scope:
        scope.span.set_tag('hello-to', hello_to)
        hello_str = format_string(hello_to)
        print_hello(hello_str)

def format_string(hello_to):
    with tracer.start_active_span('format') as scope:
        hello_str = 'Hello, %s!' % hello_to
        scope.span.log_kv({'event': 'string-format', 'value': hello_str})
        return hello_str

def print_hello(hello_str):
    with tracer.start_active_span('println') as scope:
        print(hello_str)
        scope.span.log_kv({'event': 'println'})

# main
assert len(sys.argv) == 2

tracer = init_tracer('hello-world')

hello_to = sys.argv[1]
say_hello(hello_to)

# yield to IOLoop to flush the spans
time.sleep(2)
tracer.close()
